MPI-parallel DDS with replicas

jngrad opened the pull request:
Re-implementation of #2499

Description of changes:
- the `DipolarDirectSumCpu` method was rewritten to support replicas and MPI-parallelization
   - API change: a new argument `n_replicas` was added, defaults to 0 so that old ESPResSo scripts still work
- the original `DipolarDirectSumWithReplicaCpu` method was removed
- the minimum image convention is now tested

stekajack commented:
Hi @jngrad @RudolfWeeber @fweik Did testing of the MDdS implementation vs. p3m and theory. Will write with some detail the conclusion for completeness.

The test is magnetisation of a ferrofluid, 3d pbc, \phi=4%, lambda=3. Validation against Langevin magnetisation law and modified mean-field theory (MMFT2; 2nd order). Key subtlety is the fact that direct sum and p3m use different boundary conditions (bc), therefore are not comparable as such. There are several ways to go about this. 

By default, p3m use the so-called metallic bc. MMFT2 should agree with p3m in this case. Direct sum solves for essentially a finite system (given enough replicas one can argue a pseudo-infinite system) in vacuum, which in effect means that there is implicitly a demagnetisation field in the results obtained from the direct sum. I expanded the MMFT2 to account for the demagnetisation field and should therefore be able to fit the direct sum results. Alternatively, p3m has a surface correction method implemented which is run if the \epsilon parameter is not 0. To my understanding setting the \epsilon param to 1 corresponds to p3m using vacuum boundary conditions and should therefore match MDdS  and MMFT2 with the demagnetisation field. The figure attached fits my expectations. I would say that MDdS is functioning as intended. Hope this is helpful. Cheers.



![mdds_test](https://user-images.githubusercontent.com/30287421/194884170-3d4b128d-cdc2-4537-8313-0b685a04e13e.png)

RudolfWeeber commented on src/core/magnetostatics/CMakeLists.txt:
```diff
@@ -22,7 +22,6 @@ target_sources(
           ${CMAKE_CURRENT_SOURCE_DIR}/barnes_hut_gpu.cpp
           ${CMAKE_CURRENT_SOURCE_DIR}/dds.cpp
```
pls consider renaming the file to dipolar_direct_sum or the like

jngrad replied:
done

RudolfWeeber commented on src/core/magnetostatics/dds.cpp:
```diff
@@ -30,110 +30,376 @@
+
+  /* Interaction with all the other particles */
+  for (auto pi = local_posmom_begin; pi != local_posmom_end; ++pi, ++p) {
+    // red particles
```
For interactions among local particles, each pair is only looked at once and symmetry is used to calculate the oposite force/torque. For interactions between particles at different nodes, this seems not to be the case, so f_ij and f_ji are calculated separately. the result is correct. However, in the future, this could/should be expanded to use the same symmetry and have a final reduciton of the forces. Due to the n^2 scaling, this is probably worth it. Not now, though.

RudolfWeeber commented on src/core/magnetostatics/dds.cpp:
```diff
@@ -30,110 +30,376 @@
+ */
+double
+DipolarDirectSum::long_range_energy(ParticleRange const &particles) const {
+  auto const &box_l = ::box_geo.length();
```
This kernel looks structurally different from the force kernel. Is there an mpi reduce somewhere else in Es whic accumulates the energy contributions from all the nodes?

jngrad replied:
Yes, this is managed by the `Observable_stat` framework, which is always called in parallel from the corresponding observable and analysis function.

RudolfWeeber reviewed (commented):
this will be very useful for the folks in Vienna and Stuttgart. Thanks Flo for the initial work and Jn and Deniz for finishing it.

jngrad committed:
MPI-parallel DDS with replicas

The new implementation supports MPI-parallelization and replicas.
When no replicas are used, minimum image convention is applied.
The redundant DipolarDirectSumWithReplicaCpu method was removed.

Co-authored-by: Florian Weik <fweik@icp.uni-stuttgart.de>

jngrad committed:
core: Rename source files and iterator variables

kodiakhq[bot] committed:
Merge branch 'python' into mdds_rewrite
